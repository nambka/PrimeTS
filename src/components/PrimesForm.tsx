import React from 'react';
import { sieve } from './Algo';

// Render the input form to enter two numbers
interface IProps {
  // value: number;
}
interface IState {
  inputFirstNum: string;
  inputLastNum: string;
  answer: (string | number)[];
}

class PrimesForm extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      inputFirstNum: '1',
      inputLastNum: '1000',
      answer: ['Primes']
    };
    // this.onFirstNumChange = this.onFirstNumChange.bind(this);
    // this.onLastNumChange = this.onLastNumChange.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  // onFirstNumChange(e: React.ChangeEvent<HTMLInputElement>) {
  //   this.setState({
  //     inputFirstNum: +e.target.value
  //   })
  // }

  // onLastNumChange(e: React.ChangeEvent<HTMLInputElement>) {
  //   this.setState({
  //     inputLastNum: +e.target.value
  //   })
  // }

  handleChange(e: any) {
    this.setState({
      [e.target.name]: e.target.value,
    } as IState);
  }

  handleSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    const num1 = Number(this.state.inputFirstNum)
    const num2 = Number(this.state.inputLastNum)
    const max = num1 > num2 ? num1 : num2
    const min = max === num1 ? num2 : num1

    this.setState({
      answer: sieve(max, min),
    });   
  }

  render() {
    return (
      <form id="frmFindPrimes" onSubmit={this.handleSubmit} className="container-fluid nambk-container-input">
        <h2 className="nambk-header">Prime Numbers</h2>
        <div className="nambk-header-desc">Between Two Numbers</div>
        <div>
          <input className="nambk-input" style={{width: '33%'}} 
                 type="number" min="1" max="9999999" aria-label="Input Number"
                 value={this.state.inputFirstNum} onChange={this.handleChange.bind(this)} 
                 id="inputFirstNum" name="inputFirstNum" placeholder="Enter a number"/>
          <input className="nambk-input" style={{width: '33%'}} 
                 type="number" min="1" max="9999999" aria-label="Input Number"
                 value={this.state.inputLastNum} onChange={this.handleChange.bind(this)} 
                 id="inputLastNum" name="inputLastNum" placeholder="Enter a number"/>
          <input id="btnFindPrimes" className="nambk-btn nambk-btn-primary" 
                 type="submit" value="Go"/>
        </div>
        <div id="result" className="mt-3" style={{minHeight: '4rem',}}>
          {this.state.answer.join(', ')}
        </div>   
      </form>
    );
  }
}
export default PrimesForm;