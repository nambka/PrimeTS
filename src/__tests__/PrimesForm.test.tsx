import React from 'react';
import PrimesForm from '../components/PrimesForm';
import { shallow, mount } from 'enzyme';

describe('PrimesForm: ', () => {
  let wrapper:any;
  
  beforeEach(() => {
    wrapper = shallow(<PrimesForm />);
  });

  afterEach(() => {
    jest.resetModules();
    jest.resetAllMocks();
    // wrapper.unmount();
  });

  it('should render the form and components', () => {
    const wrapper = shallow(<PrimesForm />);
    expect(wrapper.find('form').exists()).toBe(true);
    expect(wrapper.find('#inputFirstNum').length).toEqual(1);
    expect(wrapper.find('#inputLastNum').length).toEqual(1);
    expect(wrapper.find('input[type="submit"]').length).toEqual(1);
    expect(wrapper.find('#result').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  })

  it('has a title of "Prime Numbers"', () => {
    const wrapper = shallow(<PrimesForm />);
    expect(wrapper.find("h2").text()).toBe("Prime Numbers");
    expect(wrapper).toMatchSnapshot();
  });

  it('renders one form component', () => {
  const wrapper = shallow(<PrimesForm />);
  expect(wrapper.find('form')).toHaveLength(1);
  expect(wrapper).toMatchSnapshot();
  });

  it('should change the state on input1: call handleChange method', done => {
    const text = '1'
    const state = { inputFirstNum: text };
    const props = {
        fetchData: jest.fn(() => state)
    };
    const wrapper = mount(<PrimesForm {...props}/>);
    const input = wrapper.find('#inputFirstNum');

    input.props().value = text;
    input.simulate('change', { target: { value: text } });
    expect(input.get(0).props.value).toEqual(text);
    done();
  });

  it('should change the state on input2: call handleChange method', done => {
    const text = '1000'
    const state = { inputLastNum: text };
    const props = {
        fetchData: jest.fn(() => state)
    };
    const wrapper = mount(<PrimesForm {...props}/>);
    const input = wrapper.find('#inputLastNum');

    input.props().value = text;
    input.simulate('change', { target: { value: text } });
    expect(input.get(0).props.value).toEqual(text);
    done();
  });

  it('should have proper props for input1 field', () => {
    expect(wrapper.find('input[type="number"]').at(0).props()).toEqual({
      className: 'nambk-input',
      onChange: expect.any(Function),
      placeholder: 'Enter a number',
      type: 'number',
      'aria-label': 'Input Number',
      min: '1',
      max: '9999999',
      id: 'inputFirstNum',
      name: 'inputFirstNum',
      style: Object({width: '33%'}),
      value: 1,
    });
  });

  it('should have proper props for input2 field', () => {
    expect(wrapper.find('input[type="number"]').at(1).props()).toEqual({
      className: 'nambk-input',
      onChange: expect.any(Function),
      placeholder: 'Enter a number',
      type: 'number',
      'aria-label': 'Input Number',
      min: '1',
      max: '9999999',
      id: 'inputLastNum',
      name: 'inputLastNum',
      style: Object({width: '33%'}),
      value: 1000,
    });
  });

  it('should have proper props for submit button', () => {
    expect(wrapper.find('input[type="submit"]').props()).toEqual({
      className: 'nambk-btn nambk-btn-primary',
      id: 'btnFindPrimes',
      value: 'Go',
      type: 'submit',
    });
  });

  // Solution for state not updated: https://github.com/enzymejs/enzyme/issues/1999#issuecomment-491634796
  it('should change the value on input1', () => {
    wrapper.find('input[type="number"]').at(0).simulate('change', { 
      target: { name: 'inputFirstNum', value: 11 }
    });
    expect(wrapper.find('input[type="number"]').at(0).prop('value')).toEqual(11);
  });
  it('should change the value on input2', () => {
    wrapper.find('input[type="number"]').at(1).simulate('change', { 
      target: { name: 'inputLastNum', value: 22 } 
    });
    expect(wrapper.find('input[type="number"]').at(1).prop('value')).toEqual(22);
  });

  it('simulates form submit event', () => {
    const testValues = {
      inputFirstNum: '1',
      inputLastNum: '10',
      handleSubmit: jest.fn(),
    };
    const fakeEvent = { preventDefault: () => console.log('preventDefault') };
    const mockSubmit = jest.fn().mockImplementation((init) => [init, mockSubmit]);
    wrapper = shallow(<PrimesForm {...testValues} onSubmit={mockSubmit}/>);

    wrapper.find('form').simulate('submit', fakeEvent);
    expect(mockSubmit).toBeTruthy();
    // expect(testValues.handleSubmit).toHaveBeenCalledTimes(1);
    // expect(testValues.handleSubmit).toBeCalledWith({
    //     inputFirstNum: testValues.inputFirstNum, 
    //     inputLastNum: testValues.inputLastNum
    // });
  });

})