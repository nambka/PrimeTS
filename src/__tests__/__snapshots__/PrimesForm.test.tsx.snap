// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimesForm:  has a title of "Prime Numbers" 1`] = `
<form
  className="container-fluid nambk-container-input"
  id="frmFindPrimes"
  onSubmit={[Function]}
>
  <h2
    className="nambk-header"
  >
    Prime Numbers
  </h2>
  <div
    className="nambk-header-desc"
  >
    Between Two Numbers
  </div>
  <div>
    <input
      aria-label="Input Number"
      className="nambk-input"
      id="inputFirstNum"
      max="9999999"
      min="1"
      name="inputFirstNum"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "33%",
        }
      }
      type="number"
      value={1}
    />
    <input
      aria-label="Input Number"
      className="nambk-input"
      id="inputLastNum"
      max="9999999"
      min="1"
      name="inputLastNum"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "33%",
        }
      }
      type="number"
      value={1000}
    />
    <input
      className="nambk-btn nambk-btn-primary"
      id="btnFindPrimes"
      type="submit"
      value="Go"
    />
  </div>
  <div
    className="mt-3"
    id="result"
    style={
      Object {
        "minHeight": "4rem",
      }
    }
  >
    Primes
  </div>
</form>
`;

exports[`PrimesForm:  renders one form component 1`] = `
<form
  className="container-fluid nambk-container-input"
  id="frmFindPrimes"
  onSubmit={[Function]}
>
  <h2
    className="nambk-header"
  >
    Prime Numbers
  </h2>
  <div
    className="nambk-header-desc"
  >
    Between Two Numbers
  </div>
  <div>
    <input
      aria-label="Input Number"
      className="nambk-input"
      id="inputFirstNum"
      max="9999999"
      min="1"
      name="inputFirstNum"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "33%",
        }
      }
      type="number"
      value={1}
    />
    <input
      aria-label="Input Number"
      className="nambk-input"
      id="inputLastNum"
      max="9999999"
      min="1"
      name="inputLastNum"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "33%",
        }
      }
      type="number"
      value={1000}
    />
    <input
      className="nambk-btn nambk-btn-primary"
      id="btnFindPrimes"
      type="submit"
      value="Go"
    />
  </div>
  <div
    className="mt-3"
    id="result"
    style={
      Object {
        "minHeight": "4rem",
      }
    }
  >
    Primes
  </div>
</form>
`;

exports[`PrimesForm:  should render the form and components 1`] = `
<form
  className="container-fluid nambk-container-input"
  id="frmFindPrimes"
  onSubmit={[Function]}
>
  <h2
    className="nambk-header"
  >
    Prime Numbers
  </h2>
  <div
    className="nambk-header-desc"
  >
    Between Two Numbers
  </div>
  <div>
    <input
      aria-label="Input Number"
      className="nambk-input"
      id="inputFirstNum"
      max="9999999"
      min="1"
      name="inputFirstNum"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "33%",
        }
      }
      type="number"
      value={1}
    />
    <input
      aria-label="Input Number"
      className="nambk-input"
      id="inputLastNum"
      max="9999999"
      min="1"
      name="inputLastNum"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "33%",
        }
      }
      type="number"
      value={1000}
    />
    <input
      className="nambk-btn nambk-btn-primary"
      id="btnFindPrimes"
      type="submit"
      value="Go"
    />
  </div>
  <div
    className="mt-3"
    id="result"
    style={
      Object {
        "minHeight": "4rem",
      }
    }
  >
    Primes
  </div>
</form>
`;
